packages/go/index.ts

```diff
@@ -94,332 +99,454 @@ export async function build({
+    try {
+      const goModAbsPath = await findGoModPath(workPath);
+      if (goModAbsPath) {
+        debug(`Found ${goModAbsPath}"`);
       }
 
-      if (!isGoModInRootDir && config.zeroConfig && file === 'api/go.mod') {
-        // We didn't find `/go.mod` but we found `/api/go.mod` so move it to the root
+      analyzed = await getAnalyzedEntrypoint(
+        workPath,
+        entrypointAbsolute,
+        dirname(goModAbsPath)
+      );
+    } catch (err) {
+      console.log(`Failed to parse AST for "${entrypoint}"`);
+      throw err;
+    }
+
+    if (!analyzed) {
+      const err = new Error(
+        `Could not find an exported function in "${entrypoint}"
+  Learn more: https://vercel.com/docs/runtimes#official-runtimes/go
+        `
+      );
+      console.log(err.message);
+      throw err;
+    }
+
+    const parsedAnalyzed = JSON.parse(analyzed) as Analyzed;
+
+    // find `go.mod` in downloadedFiles
+    const entrypointDirname = dirname(entrypointAbsolute);
+    let isGoModExist = false;
+    let goModPath = '';
+    let isGoModInRootDir = false;
+
+    const modFileRefs = await glob('**/*.mod', workPath);
+    const modFiles = Object.keys(modFileRefs);
+
+    for (const file of modFiles) {
+      const fileDirname = dirname(file);
+      if (file === 'go.mod') {
         isGoModExist = true;
         isGoModInRootDir = true;
-        goModPath = join(fileDirname, '..');
-        const pathParts = fsPath.split(sep);
-        pathParts.pop(); // Remove go.mod
-        pathParts.pop(); // Remove api
-        pathParts.push('go.mod');
-        const newFsPath = pathParts.join(sep);
-        debug(`Moving api/go.mod to root: ${fsPath} to ${newFsPath}`);
-        await move(fsPath, newFsPath);
-        const oldSumPath = join(dirname(fsPath), 'go.sum');
-        const newSumPath = join(dirname(newFsPath), 'go.sum');
-        if (await pathExists(oldSumPath)) {
-          debug(`Moving api/go.sum to root: ${oldSumPath} to ${newSumPath}`);
-          await move(oldSumPath, newSumPath);
+        goModPath = join(workPath, fileDirname);
+      } else if (file.endsWith('go.mod')) {
+        if (entrypointDirname === fileDirname) {
+          isGoModExist = true;
+          goModPath = join(workPath, fileDirname);
+
+          debug(`Found file dirname equals entrypoint dirname: ${fileDirname}`);
+          break;
+        }
+
+        if (!isGoModInRootDir && config.zeroConfig && file === 'api/go.mod') {
+          // We didn't find `/go.mod` but we found `/api/go.mod` so move it to the root
+          isGoModExist = true;
+          isGoModInRootDir = true;
+          goModPath = join(fileDirname, '..');
+          const pathParts = file.split(sep);
+          pathParts.pop(); // Remove go.mod
+          pathParts.pop(); // Remove api
+          pathParts.push('go.mod');
+          const newFsPath = pathParts.join(sep);
+
+          debug(`Moving api/go.mod to root: ${file} to ${newFsPath}`);
+          await move(file, newFsPath);
```

Comment:
Will this work now that `newFsPath` was changed from an absolute path to a relative path? 

(Pro tip: most of the builders use fspath, to mean absolute path)

Reply:
No, that looks wrong. I thought this condition was covered by the tests, but because of the `zeroConfig` check, I don't believe it is. I'll make sure it's covered and resolve this issue.

Reply:
Oh, it happened to work because the relative path was correct, but to make sure this is true, I updated it to use an absolute path. [`7308609` (#8291)](https://github.com/vercel/vercel/pull/8291/commits/7308609dd9bbb19bea1eb224225adca669457d62)

Turns out fixture `17-go-mod-api` was already testing this case. So, no new test is necessary. The fact that this test still passes shows the change here still works.